tokio/src/runtime/io/scheduled_io.rs

```diff
@@ -458,11 +461,46 @@ impl Future for Readiness<'_> {
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         use std::sync::atomic::Ordering::SeqCst;
 
-        let (scheduled_io, state, waiter) = unsafe {
+        let (scheduled_io, state, waiter, is_waiter_registered) = unsafe {
             let me = self.get_unchecked_mut();
-            (&me.scheduled_io, &mut me.state, &me.waiter)
+            (
+                &me.scheduled_io,
+                &mut me.state,
+                &me.waiter,
+                &mut me.is_waiter_registered,
+            )
         };
 
+        if !crate::runtime::coop::has_budget_remaining() {
+            // Wasn't ready, take the lock (and check again while locked).
+            let mut waiters = scheduled_io.waiters.lock();
+
+            let w = unsafe { &mut *waiter.get() };
+
+            if *is_waiter_registered {
```

Comment:
I already mentioned this on our Discord chat, but I'll put it here too. This code isn't correct because when the coop budget is empty, you're registering the waker with the IO driver. But in this case, we need to be woken up by the coop system, and *not* by the IO waker. So you should register for readiness with the coop system instead.

Reply:
Yes. Thank you for pointing it out. I'm swamped right now with my job and university. I'll get to it as soon as I find some free time.

Reply:
No worries. That's perfectly fine.

Reply:
@maminrayej I can take this over from you if you're busy, I'd like to get this change over the finish line relatively quickly.

Reply:
@Noah-Kennedy Absolutely, feel free to take it over. Thanks for stepping in.

Reply:
no problem!